### PR TITLE
charts/vault-operator: Allow specifying rather than generating certs

### PR DESCRIPTION
**Pull request by donaldguy**
While its a ~cluster-internal concern, I was not incredibly comfortable
letting these certs be rotated on every `helm upgrade`.

As it stands w/o this option, an upgrade is a cluster breaking change without `--recreate-pods` (at least while we live in a world without livenessProbes). And ~rolling upgrades are currently impossible (as is doing an upgrade to e.g. just tweak monitoring options)

It also undermines the utility of tools like https://github.com/databus23/helm-diff

I think this is a good option to have, if people want to put in the ground work (or just e.g. use the provided terraform)

### CODE REVIEW

**Comment by donaldguy** [comment]
Note that my testing doesn't suggest that the current vault-operator chart on the `HEAD` of master is exactly functional as it is (following the API review changes perhaps?) and the e2e seems to only apply/test the 0.2.0 chart

But I did cherry-pick this over to the `0.2.0` tag (see here: https://github.com/donaldguy/operator/tree/PR238-on-0.2.0 ) and test from there – both with the defaults, with a values.yaml, and with my provided Terraform (although admittedly on Helm 3 using a build off terraform-providers/terraform-provider-helm#378)

**Comment by donaldguy** [comment]
Re:
>Note that my testing doesn't suggest that the current vault-operator chart on the HEAD of master is exactly functional as it is

in both cases (with and without this PR) - pods fail 
```
Error: timed out waiting for CRD: postgresroles.authorization.kubedb.com is forbidden: User "system:serviceaccount:kube-system:vault-operator" cannot list resource "postgresroles" in API group "authorization.kubedb.com" at the cluster scope
```

(as a result of 26d38547bd174334fdcb60428892faeb4c95db3e or something else in #204  ?) 

Maybe this is only an issue when you run with :0.2.0 images and the current chart ...

**Comment by donaldguy** [comment]
yeah ^ I am seeing the above error in the logs when running `make ct` on master as well

but its "passing" I don't think that test provides any signal with `apiserver.healtcheck.enabled: false` ... but you probably no that. 

But anyway, this PR doesn't break it :-)

File: charts/vault-operator/values.yaml [hunk 1]
```diff
@@ -97,6 +97,14 @@ apiserver:
   # healthcheck configures the readiness and liveliness probes for the operator pod.
   healthcheck:
     enabled: false
+  k8s2operatorCerts:
```
Comment (tamalsaha) [inline charts/vault-operator/values.yaml]: Nit: `k8s2operatorCerts` -> `servingCerts`

**Comment by tamalsaha** [comment]
Thanks for the pr. I have wanted to support this use-case where certs come from outside. Can you rebase the pr to fix conflicts? I removed the rbac.create value.

Regarding testing the chart, we are looking to cut a new release as the documentation is ready.

**Comment by donaldguy** [comment]
rebased :-)

File: charts/vault-operator/templates/apiregistration.yaml [hunk 2]
```diff
@@ -1,8 +1,21 @@
+{{- $caCrt := "" }}
+{{- $serverCrt := "" }}
+{{- $serverKey := "" }}
+{{- if .Values.apiserver.servingCerts.generate }}
 {{- $ca := genCA "ca" 3650 }}
 {{- $cn := include "vault-operator.fullname" . -}}
 {{- $altName1 := printf "%s.%s" $cn .Release.Namespace }}
 {{- $altName2 := printf "%s.%s.svc" $cn .Release.Namespace }}
-{{- $cert := genSignedCert $cn nil (list $altName1 $altName2) 3650 $ca }}
+{{- $altName3 := printf "%s.%s.svc.cluster.local" $cn .Release.Namespace }}
```
Comment (tamalsaha) [inline charts/vault-operator/templates/apiregistration.yaml]: `.cluster.local` is configurable by the `--cluster-domain` flag to kubelet. https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/ . I think we should to hardcode this here.

File: charts/vault-operator/example-terraform.tf [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  private_key_pem = tls_private_key.kubevault_server.private_key_pem
+
+  subject {
+    common_name = local.kubevault_release_fullname
+  }
+
+  dns_names = [
+    "${local.kubevault_release_fullname}.${local.kubevault_namespace}",
+    "${local.kubevault_release_fullname}.${local.kubevault_namespace}.svc",
+    "${local.kubevault_release_fullname}.${local.kubevault_namespace}.svc.cluster.local"
```
Comment (tamalsaha) [inline charts/vault-operator/example-terraform.tf]: Also here?

Comment (donaldguy) [reply charts/vault-operator/example-terraform.tf]: If you insist - I didn't forget  – I was thinking like "well if you are creating the certs for the chart, you are taking on the responsibility of making sure its right";

but sure. one sec.

File: charts/vault-operator/example-terraform.tf [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  validity_period_hours = 87600
+  set_subject_key_id    = false
+
+  is_ca_certificate = true
+  allowed_uses = [
+    "digital_signature",
+    "key_encipherment",
+    "cert_signing",
+    "server_auth",
+    "client_auth"
```
Comment (tamalsaha) [inline charts/vault-operator/example-terraform.tf]: I think these are not needed for CA cert.

File: charts/vault-operator/example-terraform.tf [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  cert_request_pem = tls_cert_request.kubevault_server.cert_request_pem
+
+  validity_period_hours = 87600
+  set_subject_key_id    = false
+
+  allowed_uses = [
+    "digital_signature",
+    "key_encipherment",
+    "server_auth",
+    "client_auth"
```
Comment (tamalsaha) [inline charts/vault-operator/example-terraform.tf]: I think `"client_auth"` is not needed for serving cert.

Comment (donaldguy) [reply charts/vault-operator/example-terraform.tf]: to this and the above, probably not – but I just replicated what the sprig helpers did exactly to minimize change testing.

Comment (donaldguy) [reply charts/vault-operator/example-terraform.tf]: I notably knowing did _not_ follow exactly `onessl` - as it also adds an IP:127.0.0.1 to certs

Comment (tamalsaha) [reply charts/vault-operator/example-terraform.tf]: If you followed the sprig output, it should be ok.

**Pull request merged**